src/reduce.jl

```diff
@@ -0,0 +1,98 @@
+export @groupreduce, @subgroupreduce
+
+"""
+
+@subgroupreduce(op, val)
+
+reduce values across a subgroup. This operation is only supported if subgroups are supported by the backend.
+"""
+macro subgroupreduce(op, val)
+    quote
+        $__subgroupreduce($(esc(op)),$(esc(val)))
+    end
+end
+
+function __subgroupreduce(op, val)
+    error("@subgroupreduce used outside kernel, not captured, or not supported")
+end
+
+"""
+
+@groupreduce(op, val, neutral, use_subgroups)
+
+Reduce values across a block
+- `op`: the operator of the reduction
+- `val`: value that each thread contibutes to the values that need to be reduced
+- `netral`: value of the operator, so that `op(netural, neutral) = neutral``
+- `use_subgroups`: make use of the subgroupreduction of the groupreduction
```

Comment:
I see the value of having a common implementation.

So I would define:

```
@reduce(op, val, neutral)
__reduce(op, val, neutral)
```

And then maybe:

```
__subgroupreduce & __subgroupsize (No @ version)
__can_subgroup_reduce(T) = false
```

And then we could define:

```
__reduce(__ctx___, op, val, neutral, ::Val{true})
__reduce(__ctx___, op, val, neutral, ::Val{false})
```

as you have here.

```
function __reduce(op, val::T, neutral::T) where T
     __reduce(op, val, neutral, Val(__can_subgroup_reduce(T)))
end
```